### PR TITLE
#5560: Add `all_reduce` op (experimental)

### PR DESCRIPTION
**Pull request by Aswinmcw**
Tracking Issue : #5560 

Add `all_reduce` op under experimental folder

## Q: Is only sum as reduce op is supported?
 
A: Today yes. Future, no. We'd need to support min, max, and other reduction operations.

## Q: What is the output shape after all_reduce, is it equal to the shape on one device?

A: Yes (for cases where for whatever reason one of the devices may have less data, it would be the size of the larger tensors, and the device with a smaller tensor would logically be padded with identity value elements. Whether this is a legitimate use case for this operation I am not sure atm).

## Q: After all reduce, will all devices have the same reduced tensor?

A: Yes

## Q: Do we do reduce over the batch dimension only?

A: Not exactly. All-reduce is a rank-wise operation. That is we are doing element-wise reductions for tensors across devices. For reference, we are following the semantics outlined [here](https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/usage/collectives.html#allreduce).

## Q: What is the memory footprint on one device? e.g we have 32 devices and call reduce on tensor with volume 16k how how many dram will be allocated during this op?

A:
#### Today
memory consumption is inefficient because we are providing the least optimized functional implementation for all-reduce (composite all-gather + eltwise operation).

if # devices = n and input tensor size per device = m, then the total memory working set size (per chip) will be (today) (n + 1) * m

That is n * m needed to store the intermediate all-gather results and then another m to store the final output. After the operation the n *m can be returned to the allocation

#### Near future
We will enable the reduce_scatter + all-gather (composite) optimization that will both reduce memory pressure and improve performance. This has 2 constraints:

All-gather and reduce scatter must enable support for non-uniform tensor sizes across chips
the input tensor shape must have a dim size > n
technically it just needs a volume > n but that would require multi-dimension scatter/concat dim for reduce scatter/all-gather which is totally out of the picture for the foreseeable future
With this version total memory pressure (per chip) would be m + m * (1/n) where m * (1/n) is the size of the intermediate tensor between reduce_scatter and all-gather

#### Long Term
We will support a proper non-composite implementation in which the memory needed per chip will simply b m

### Checklist
- [x] Post commit CI passes - https://github.com/tenstorrent/tt-metal/actions/runs/11363163144
- [x] T3K Pipelines - https://github.com/tenstorrent/tt-metal/actions/runs/11363166696
- [ ] Blackhole Post commit (if applicable)
- [ ] Model regression CI testing passes (if applicable)
- [ ] Device performance regression CI testing passes (if applicable)
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp:62]: please remove the stale dim arg

File: tests/ttnn/unit_tests/operations/test_all_reduce.py
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_reduce.py:57]: let's avoid default valuing the `num_workers` and `num_buffers_per_channel` to avoid the test diverging from the implementation. The implementation should be free to choose with support for overridden values as best effort support.

File: tests/ttnn/unit_tests/operations/test_all_reduce.py
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_reduce.py:1]: Please rename to denote the target test list (e.g. `test_all_reduce_t3000_frequent.py`)

File: tests/ttnn/unit_tests/operations/test_all_reduce.py
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_reduce.py:75]: For the scope of this function, please rename to `mesh_device`

File: tests/ttnn/unit_tests/operations/test_all_reduce.py
Comment (SeanNijjar) [inline tests/ttnn/unit_tests/operations/test_all_reduce.py:219]: Let's default to 2 iterations. I want to start moving all CCL op tests to default to > 1 iteration.

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp:87]: also please remove dim arg here.

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp
Comment (SeanNijjar) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/all_reduce_pybind.cpp:66]: Random thought (not specific feedback to this PR), but maybe we can have "standard" descriptions for some of these standard args and just include them as a part of the formatted string...? This should really help us have uniform descriptions!

**Review by SeanNijjar — APPROVED** [review]
Thanks a lot for this quick turnaround! Mostly minor requests for change, however, I didn't see the tests added to the t3000 frequent tests list - can you please add the tests there?

For future work we can add the optimized (reduce_scatter + all_gather) composite and enable sharding support (line reduce scatter sharding support is now in PR).

Please rerun the frequent pipeline and ensure you see the tests there. 

I'm approving now to compensate for time difference but please apply these changes and get a new clean run before merging. Thanks a lot!

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:76]: Do we really have to use launch_op here? 
Is that because the output type is vector of tensors?

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:94]: @SeanNijjar is this code repeated in one way or another across ccl ops?

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (ayerofieiev-tt) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:116]: Is this a view or actual data shuffling will be needed?

**Review by ayerofieiev-tt — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:116]: We are adding a temporary outer dim to perform all_gather operation on that dim, so it is actual data shuffling needed

File: tests/ttnn/unit_tests/operations/test_all_reduce_t3000_frequent.py
Comment (Denys88) [inline tests/ttnn/unit_tests/operations/test_all_reduce_t3000_frequent.py:201]: Are there any other ops?

File: tests/ttnn/unit_tests/operations/test_all_reduce_t3000_frequent.py
Comment (Denys88) [inline tests/ttnn/unit_tests/operations/test_all_reduce_t3000_frequent.py:171]: Please add tests for all dimensions including the 0 and 1 too.
Please add tests for non-divisible by 32 dimensions or explicitly write that it is not supported.

**Review by Denys88 — COMMENTED** [review]
HI @Aswinmcw 
A few questions:
1) does it mean only sum as reduce op is supported?
2) What is the output shape after all_reduce, is it equal to shape on one device?
3) Right now after all reduce all devices have the same reduced tensor?
4) Do we do reduce over the batch dimmension only? 
5) What is the memory footprint on one device? e.g we have 32 devices and call reduce on tensor with volume 16k how how many dram will be allocated during this op?

**Comment by Aswinmcw** [comment]
> HI @Aswinmcw A few questions:
> 
> 1. does it mean only sum as reduce op is supported?
> 2. What is the output shape after all_reduce, is it equal to shape on one device?
> 3. Right now after all reduce all devices have the same reduced tensor?
> 4. Do we do reduce over the batch dimmension only?
> 5. What is the memory footprint on one device? e.g we have 32 devices and call reduce on tensor with volume 16k how how many dram will be allocated during this op?

Hi @Denys88 , answering a few questions,

- Working of this op is individual different input tensors will be created and stored on each devices then a new temporary outer dim will be created and the remaining dims will be merged to maintain the 4D structure and this all_gather operation will happen only on the newly created 0th dim
- Then it will reduce all gathered tensors and finally all devices will have same data
- After all_reduce operation, the output shape will be same as the shape of a single chip

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:94]: Yes. I will commonize.

File: tests/ttnn/unit_tests/operations/test_all_reduce_t3000_frequent.py
Comment (SeanNijjar) [reply tests/ttnn/unit_tests/operations/test_all_reduce_t3000_frequent.py:206]: For full generality yes but to start no. The code will throw of the user tries a reduction that isn't sum.

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:116]: If we are just inserting am outer dim though I would hope under the hood this just resolves to a reinterpret

**Comment by SeanNijjar** [comment]
> HI @Aswinmcw A few questions:
> 
> 1. does it mean only sum as reduce op is supported?
Today yes. Future, no. We'd need to support min, max, and other reduction operations.

> 2. What is the output shape after all_reduce, is it equal to shape on one device?
Yes (for cases where for whatever reason one of the devices may have less data, it would be the size of the larger tensors, and the device with a smaller tensor would logically be padded with identity value elements. Whether this is a legitimate use case for this operation I am not sure atm).

> 3. Right now after all reduce all devices have the same reduced tensor?
Yes

> 4. Do we do reduce over the batch dimmension only?
Not really. All-reduce is a rank-wise operation. That is we are doing element-wise reductions for tensors across devices. 

> 5. What is the memory footprint on one device? e.g we have 32 devices and call reduce on tensor with volume 16k how how many dram will be allocated during this op?

#### Today
memory consumption is inefficient because we are providing the least optimized functional implementation for all-reduce (composite all-gather + eltwise operation). 

if `# devices = n` and `input tensor size per device = m`, then the total memory working set size (per chip) will be (today) `(n + 1) * m`
- That is n * m needed to store the intermediate all-gather results and then another m to store the final output. After the operation the n *m can be returned to the allocation

#### Near future
We will enable the reduce_scatter + all-gather (composite) optimization that will both reduce memory pressure and improve performance. This has 2 constraints:
1) All-gather and reduce scatter must enable support for non-uniform tensor sizes across chips
2) the input tensor shape must have a dim size > n
  - technically it just needs a volume > n but that would require multi-dimension scatter/concat dim for reduce scatter/all-gather which is totally out of the picture for the foreseeable future

With this version total memory pressure (per chip) would be `m + m * (1/n)` where `m * (1/n)` is the size of the intermediate tensor between reduce_scatter and all-gather

#### Long Term
We will support a proper non-composite implementation in which the memory needed per chip will simply b `m`

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:111]: should we check whether rank >= 3? 
Another option might be `for (uint32_t i = 0; i + 3 <= rank; ++i)`

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:107]: wdyt about using `input_tensor.get_logical_shape()` to clearly represent the intention?

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:111]: Yes, I will add a check before this

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (Aswinmcw) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:107]: I will check and if that works as expected, i will update it

File: ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp
Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:111]: We want to merge all of the outer or batch dims (0 to n-3 inclusive) so we can insert a new dimension and sum along that new other dimension. Though that being said there is a bug here for rank < 2 and for that we should fix how `new_shape` is calculated.

Comment (SeanNijjar) [reply ttnn/cpp/ttnn/operations/experimental/ccl/all_reduce/device/all_reduce_op.cpp:111]: Oh I see what you were referring to @rfurko-tt - `rank` is unsigned. Then disregard the first part of my comment as I thought you were alluding to something else. The latter part still applies here.

**Review by ttmchiou — APPROVED** [review]

**Pull request merged**
